internal/terraform/context_apply2_test.go

```diff
@@ -2637,6 +2638,489 @@ resource "test_object" "foo" {
+			return providers.PlanResourceChangeResponse{
+				PlannedState: req.ProposedNewState,
+			}
+		},
+		ApplyResourceChangeFn: func(req providers.ApplyResourceChangeRequest) providers.ApplyResourceChangeResponse {
+			mapKey := req.Config.GetAttr("name").AsString() // should always be known during apply
+			appliedVals.Store(mapKey, req.PlannedState)
+			return providers.ApplyResourceChangeResponse{
+				NewState: req.PlannedState,
+			}
+		},
+	}
+	state := states.NewState()
+	reset := func(newState *states.State) {
+		if newState != nil {
+			state = newState
+		}
+		plannedVals = sync.Map{}
+		appliedVals = sync.Map{}
+	}
+	normalMap := func(m *sync.Map) map[any]any {
+		ret := make(map[any]any)
+		m.Range(func(key, value any) bool {
+			ret[key] = value
+			return true
+		})
+		return ret
+	}
+	resourceInstancesActionsInPlan := func(p *plans.Plan) map[string]plans.Action {
+		ret := make(map[string]plans.Action)
+		for _, cs := range p.Changes.Resources {
+			// Anything that was deferred will not appear in the result at
+			// all. Non-deferred actions that don't actually need to do anything
+			// _will_ appear, but with action set to [plans.NoOp].
+			ret[cs.Addr.String()] = cs.Action
+		}
+		return ret
+	}
+	outputValsInState := func(s *states.State) map[string]cty.Value {
+		ret := make(map[string]cty.Value)
+		for n, v := range s.RootOutputValues {
+			ret[n] = v.Value
+		}
+		return ret
+	}
+	cmpOpts := cmp.Options{
+		ctydebug.CmpOptions,
+	}
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+	})
+	reset(states.NewState())
+
+	// Round 1: Should plan and apply changes only for test.a
+	{
+		t.Logf("--- First plan: var.each is unknown ---")
+		plan, diags := ctx.Plan(cfg, state, &PlanOpts{
+			Mode: plans.NormalMode,
+			SetVariables: InputValues{
+				"each": {
+					Value:      cty.DynamicVal,
+					SourceType: ValueFromCaller,
+				},
+			},
+		})
+		assertNoDiagnostics(t, diags)
+		checkPlanApplyable(t, plan)
+		if plan.Complete {
+			t.Fatal("plan is complete; should have deferred actions")
+		}
+
+		gotPlanned := normalMap(&plannedVals)
+		wantPlanned := map[any]any{
+			// We should've still got one plan request for each resource block,
+			// but the one for "b" doesn't yet know its own name, so is
+			// captured under "<unknown>".
+			"a": cty.ObjectVal(map[string]cty.Value{
+				"name":           cty.StringVal("a"),
+				"upstream_names": cty.NullVal(cty.Set(cty.String)),
+			}),
+			"<unknown>": cty.ObjectVal(map[string]cty.Value{
+				// this one is the placeholder for all possible instances of test.b,
+				// describing only what they all have in common.
+				// Terraform knows at least that all of the names start
+				// with "b:", even though the rest of the name is unknown.
+				"name": cty.UnknownVal(cty.String).Refine().
+					StringPrefixFull("b:").
+					NotNull().
```

Comment:
This is a neat idea for a test case! I had forgotten about the helpful overlap between refinements and deferred actions so seeing it called out here is great.

Reply:
Honestly I didn't actually explicitly intend to capture this here... it just emerged naturally from how I constructed the test fixture! But I did quite like to see it once I saw it, so I kept it. :grinning: